### PR TITLE
Add terraform ELMS module - first iteration

### PR DESCRIPTION
**Pull request by MagdaPaj**
First iteration of the ELMS terraform module. That includes:
* add dev container setup
* add only the ELMS terraform module and delete edge device twin tag dependency
* do not use default Event Hub Consumer group
* add conditional Event Hub creation
* create terraform environment with only the ELMS module
* add documentation

### CODE REVIEW

File: terraform/environment/main.tf [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+  }
+
+  backend "azurerm" {
+  }
+}
+
+provider "azurerm" {
+  features {}
+  #skip_provider_registration = true
```
Comment (Ioana37) [inline terraform/environment/main.tf]: This was needed in the past when a user does not have enough access permissions to the resource group. Should we delete it?

**Review by Ioana37 — APPROVED** [review]

File: terraform/environment/main.tf (hunk 1 above)
Comment (MagdaPaj) [reply terraform/environment/main.tf]: I think so.

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.140.1/containers/dotnetcore/.devcontainer/base.Dockerfile
```
Comment (MagdaPaj) [inline .devcontainer/Dockerfile:1]: I tried with creating DevContainer from VSCode and I tried with Ubuntu and just adding terraform and azure-cli. Then the dockerfile is way simpler:
```
# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.202.5/containers/ubuntu/.devcontainer/base.Dockerfile

# [Choice] Ubuntu version (use hirsuite or bionic on local arm64/Apple Silicon): hirsute, focal, bionic
ARG VARIANT="hirsute"
FROM mcr.microsoft.com/vscode/devcontainers/base:0-${VARIANT}

# [Optional] Uncomment this section to install additional OS packages.
# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
#     && apt-get -y install --no-install-recommends <your-package-list-here>
```

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
```
Comment (MagdaPaj) [inline .devcontainer/devcontainer.json:1]: And devcontainer.json.

```
// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
// https://github.com/microsoft/vscode-dev-containers/tree/v0.202.5/containers/ubuntu
{
	"name": "Ubuntu",
	"runArgs": ["--init"],
	"build": {
		"dockerfile": "Dockerfile",
		// Update 'VARIANT' to pick an Ubuntu version: hirsute, focal, bionic
		// Use hirsute or bionic on local arm64/Apple Silicon.
		"args": { "VARIANT": "focal" }
	},

	// Set *default* container specific settings.json values on container create.
	"settings": {},


	// Add the IDs of extensions you want installed when the container is created.
	"extensions": [],

	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	// "forwardPorts": [],

	// Use 'postCreateCommand' to run commands after the container is created.
	// "postCreateCommand": "uname -a",

	// Comment out connect as root instead. More info: https://aka.ms/vscode-remote/containers/non-root.
	"remoteUser": "vscode",
	"features": {
		"terraform": "latest",
		"azure-cli": "latest"
	}
}
```

So I'm not sure if we want to change it now, or later? And not sure which devcontaienr we should use. There are many options (C#, etc)
image

File: terraform/modules/iotelms/main.tf [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+resource "azurerm_iothub_endpoint_eventhub" "elms" {
+  count               = var.send_metrics_device_to_cloud == true ? 1 : 0
+  resource_group_name = var.rg_name
+  iothub_name         = var.iothub_name
+  name                = "metricscollector-${var.name_identifier}"
+
+  connection_string = azurerm_eventhub_authorization_rule.elms[0].primary_connection_string
+}
+
+resource "azurerm_iothub_route" "elms" {
```
Comment (MagdaPaj) [inline terraform/modules/iotelms/main.tf:233]: we had a brief discussion about it. But let's get more feedback.

For the moment we are modifying IoT Hub that customer is passing as a variable. And creating the route maybe a bit dangerous because of this:
" When you create new endpoints and routes, messages stop flowing to the built-in endpoint unless you create a separate route and direct them there. If no routes to the built-in endpoint exist, enabling a fallback route will direct any messages that don't match a route query to that endpoint"

So basically if a customer relies on built-in endpoint, we will break it by provisioning the ELMS. 
I think that PowerShell has the same behavior. But it still seems to me a bit dangerous. Options that I see:
- we can leave as-is and document it clearly, so they know what we will do with their IoT Hub
- we can delete it, and just ask customer to modify it manually (for example with Azure CLI commands).

What do you think? We can also do it later, not part of this PR.

File: docs/provision-elms-with-terraform.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+## 2. Log into Azure CLI
+
+Open a terminal, run the following commands and, if needed, select the Azure subscription which will be targeted by the Terraform deployment:
+
+```shell
+az login
+az account set --subscription="<subscription_id>"
+```
+
+## 3. Create the Terraform state storage
```
Comment (Salazander) [inline docs/provision-elms-with-terraform.md:31]: I think it's important to note here that using a backend state storage is optional.  If the user doesn't want to manage the Azure resources with Terraform beyond their creation, they can also just run the commands locally without any remote storage

File: terraform/environment/main.tf (hunk 1 above)
Comment (Salazander) [reply terraform/environment/main.tf]: Issues with provider registration is something first-time users will face. It relates to:
https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers#registration

File: terraform/modules/iotelms/main.tf (hunk 4 above)
Comment (Salazander) [reply terraform/modules/iotelms/main.tf:233]: How about enabling the fallback route through terraform?

"If there are no routes to the built-in-endpoint and a fallback route is enabled, only messages that don't match any query conditions on routes will be sent to the built-in-endpoint."

If that's not working out, I would just document the behavior.

Comment (MagdaPaj) [reply terraform/modules/iotelms/main.tf:233]: Hm... I was wrong. The docs quote I took from the Azure Portal IoT Hub. But there is [a better documentation](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-d2c#built-in-endpoint-as-a-routing-endpoint).

>So basically if a customer relies on built-in endpoint, we will break it by provisioning the ELMS.

We won't. If they rely on built-in endpoint, they must have fallback route enabled. So once we deploy the ELMS with a new route, we won't break anything.

So I don't know if we should update the fallback route. Also I don't know how terraform will behave if there is already a fallback route enabled, and we try to enable it again through the terraform. So we may need to have another conditional resource creation (that complicates the code). But that needs to be validated.

My concern was more about modifying IoT Hub from terraform. And about breaking its functionality. It is a resource not created by us. And maybe a user won't notice that we will modify it from terraform. So that's why I was thinking about asking user to explicitly modify it.
But I was basically wrong, and now I don't see what we could break with it.

Comment (vjrantal) [reply terraform/modules/iotelms/main.tf:233]: I think it is good that we think carefully how we are changing the existing resource. Adding the fallback route should maintain the functionality in cases where customer read from the built-in endpoint. In case the user has existing custom routes and have explicitly wanted to disable fallback, the change coming through terraform might be unwanted.

Perhaps in the readme, highlight that the terraform plan should be reviewed before applying to a resource that is being used?

Comment (Ioana37) [reply terraform/modules/iotelms/main.tf:233]: Agreed in the daily to add the fallback route and document it.

File: docs/provision-elms-with-terraform.md (hunk 5 above)
Comment (Ioana37) [reply docs/provision-elms-with-terraform.md:31]: We will document that it is optional.

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (Ioana37) [reply .devcontainer/devcontainer.json:1]: Improvements to the dev container will be handled on a separate branch.

File: terraform/environment/main.tf (hunk 1 above)
Comment (Ioana37) [reply terraform/environment/main.tf]: No service principal creation is required in the ELMS files so we will remove the provider registration.

File: terraform/environment/main.tf [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  rg_name                      = var.rg_name
+  location                     = var.location
+  name_identifier              = var.name_identifier
+  random_id                    = lower(random_id.elms.hex)
+  iothub_id                    = var.iothub_id
+  iothub_name                  = var.iothub_name
+  send_metrics_device_to_cloud = var.send_metrics_device_to_cloud
+}
+
+module "iotedgevm" {
```
Comment (MagdaPaj) [inline terraform/environment/main.tf]: I guess that this shouldn't be here, right?

File: terraform/modules/iotelms/main.tf (hunk 4 above)
Comment (MagdaPaj) [reply terraform/modules/iotelms/main.tf:233]: I think Azure documentation is a bit unclear. I'll try to write it, so then maybe I'll understand it better.

**Case 1: only built-in endpoint used, no routes** 
From [here](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-d2c#built-in-endpoint-as-a-routing-endpoint): "Even if no routes are created, a fallback route must be enabled to route messages to the built-in endpoint." So the fallback route is enabled. And if we add a new route, then everything will work ✅
**Case 2: routes defined, fallback disabled**
We will just add a new route, we won't change other routes. So I think everything will work also. In this setup I guess that a user wants to ignore some of the messages. Fallback is not needed, customer relies on all defined routes. But we won't change this functionality. So all good ✅

Assuming that my understanding and Azure documentation is correct, we should be fine. And we shouldn't create a fallback route. And, as @machteldbogels checked, it it not possible 
image


So I would just mentioned something like this in the documentation:

This command has an impact on the pre-existent IoT Hub. It adds a new IoT Hub endpoint and route. Make sure to carefully review the Terraform plan and that there is no impact to your existing IoT Hub message routing before agreeing to the changes.

File: terraform/environment/main.tf (hunk 6 above)
Comment (MagdaPaj) [reply terraform/environment/main.tf]: Deleted
